File: Leetcode/79.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function (board, word) {
    if (!board.length || !word.length) return false;
    word = word.split('');
    const row = board.length,
        col = board[0].length;
    let visited = Array.from(Array(row), () => new Array(col).fill(0));
    for (let i = 0; i < row; i++) {
        for (let j = 0; j < col; j++) {
            if (dfs(board, i, j, word, visited, 0)) {
                return true;
            }
        }
    }
    return false;
};

function dfs(board, row, col, word, visited, count) {
    if (count === word.length) return true;
    if (row < 0 || row >= board.length || col < 0 || col >= board[0].length || board[row][col] !== word[count]) { return false; }
    if (visited[row][col] === 0) {
        visited[row][col] = 1;
        if (dfs(board, row + 1, col, word, visited, count + 1) ||
            dfs(board, row, col + 1, word, visited, count + 1) ||
            dfs(board, row - 1, col, word, visited, count + 1) ||
            dfs(board, row, col - 1, word, visited, count + 1)) {
            return true;
        }
        visited[row][col] = 0;
    }
    return false;
}

File: Leetcode/79.java
class Solution {
    public boolean find(char[][] board, boolean[][] matrix, int i, int j, int row, int col, String word){
        if (word.length() == 0) return true;
        if (i < 0 || j < 0 || i >= row || j >= col || matrix[i][j] == true || board[i][j] != word.charAt(0)){
            return false; 
        }
        matrix[i][j] = true;
        String temp = word.substring(1);
        if (
            find(board, matrix, i + 1, j, row, col, temp) ||
            find(board, matrix, i - 1, j, row, col, temp) ||
            find(board, matrix, i, j + 1, row, col, temp) ||
            find(board, matrix, i, j - 1, row, col, temp) 
        ){
            return true;
        }
        matrix[i][j] = false;
        return false;
    }

    public boolean exist(char[][] board, String word) {
        int row = board.length, col = board[0].length;
        
        boolean[][] matrix = new boolean[row][col];
        for (int i = 0; i < row; i++){
            for(int j = 0; j < col; j++){
                if (board[i][j] == word.charAt(0)){
                    if(find(board, matrix, i, j, row, col, word)){
                        return true;
                    }
                }
            } 
        }
        return false;
    }
    
}